### PR TITLE
Streaming Conv 2D and Batchnorm 2D implementations

### PR DESCRIPTION
**Pull request by DamRsn**
Hi Jatin, 

I have implemented two new layers, 2d streaming convolution and 2d batch normalization, for a project that I am currently working on for the Neural Audio Plugin Competition. These layers are designed specifically for neural networks that process frequency domain data or similar. When those layers are used, the model runs on every frame, which contains a certain number of frequency bins, instead of every individual sample.

At the moment, I have only implemented these layers using the Eigen backend. 

Please find below some details of the implementation. Let me know what you think!

- Implementation of streaming 2D convolutions with Eigen (templated and non-templated)
  - In Tensorflow, the dimension should follow N x T x F x C (Batch, Time, Feature, Channel)
  - Support for "same" and "valid" padding (works the same way as tensorflow)
  - Support for stride on feature axis
  - Support for dilation on time axis
 
The chosen streaming approach works by performing all necessary calculations involving a frame as soon as it is acquired, and then storing the obtained results in the layer states. These states are utilized to generate the accurate output.

- Implementation of Batchnorm 2D with Eigen. Works very similarly as batchnorm1d, but now the channels consist of more than 1 value. But the number of weights stays the same. Only working when `axis=-1` in TF/Keras.

- Minimal changes to original API:
  - `in_size` and `out_size` are still used but set as `num_filters_in * num_features_in` and `num_filters_out * num_features_out` respectively.

- Test coverage for new layers, similar to other tests with some minor modifications to deal with frame alignment issues with different kind of paddings.

### CODE REVIEW

**Comment by jatinchowdhury18** [comment]
Thanks for the PR, at a glance these changes look great! Since there's a lot of changes here, it's going to take me a minute to review them thoroughly, but I just wanted to let you know that I am looking at it.

Supporting only an Eigen implementation is fine for now, but in order to stay compatible with the other backends, it would probably make sense to add some `#if RTNEURAL_USE_EIGEN` gaurds to the "main" headers (e.g. `batchnorm2d.h`), and for the relevant tests. That should also help the CI runs to pass.

**Comment by DamRsn** [comment]
Cool thanks!

I've just added a few `#if RTNEURAL_USE_EIGEN` so that other backend can compile

File: RTNeural/conv2d/conv2d_eigen.tpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+template <typename T>
+Conv2D<T>& Conv2D<T>::operator=(const Conv2D& other)
+{
+    return *this = Conv2D<T>(other);
+}
+
+template <typename T>
+void Conv2D<T>::setWeights(const std::vector<std::vector<std::vector<std::vector<T>>>>& inWeights)
+{
+    conv1dLayers.resize(kernel_size_time, Conv1DStateless<T>(num_filters_in, num_features_in, num_filters_out, kernel_size_feature, stride, valid_pad));
```
Comment (jatinchowdhury18) [inline RTNeural/conv2d/conv2d_eigen.tpp:49]: Would it be possible to move this to the layer constructor? It's not really a strict requirement, but I've been trying to keep all of the layer's memory allocation restricted to just the constructor.

File: python/conv2d.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+n_frames = 50
+n_features = 23
+
+# construct TensorFlow model
+model = keras.Sequential()
+model.add(keras.layers.InputLayer(input_shape=(n_frames, n_features, 1)))
+model.add(keras.layers.Conv2D(2, (5, 5), dilation_rate=(2, 1), strides=(1, 1), padding='valid', kernel_initializer='random_normal', bias_initializer='random_normal', activation='relu'))
+model.add(keras.layers.Conv2D(3, (4, 3), dilation_rate=(1, 1), strides=(1, 2), padding='same', kernel_initializer='random_normal', bias_initializer='random_normal'))
+model.add(keras.layers.BatchNormalization(momentum=0.0, epsilon=0.01, beta_initializer='random_normal', gamma_initializer='glorot_uniform', moving_mean_initializer="random_normal", moving_variance_initializer="ones"))
```
Comment (jatinchowdhury18) [inline python/conv2d.py:18]: Would it be possible to add another 2D BatchNorm layer, just to test the "affine" code path?

**Review by jatinchowdhury18 — COMMENTED** [review]
Cool, I had a couple of little comments, but things are looking great overall!

One thing that might be worth thinking about is the best way to document the input/output data layout for the 2D layers. The test case that you wrote is a good example, but I'm just trying to think of how to make it most obvious to the user. Probably the right answer is something like:
- A section in the README or other text file decribing the input/output layout
- Have the doc-strings for the relevant layers point to the README/text file
- Have a dedicated example for 2D networks

Anyway, the documentation stuff can be cleaned up after merging this PR if that's more convenient, just to avoid blocking ourselves.

Looks like the CI jobs are mostly passing as well... For the "Auto-Format" job, running `clang-tidy` on your changes would probably fix that one, but no worries if you aren't able to. The code coverage stuff looks okay as well (most of the new code paths that aren't being tested are some of the safety checks on model_loader.h). so don't worry that those are still showing up red.

File: tests/conv2d_model.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    return total_pad_left;
+}
+
+void processModelNonT(RTNeural::Model<TestType>& model, const std::vector<TestType>& xData, std::vector<TestType>& yData, int num_frames, int num_features_in, int num_features_out)
+{
+    model.reset();
+
+    TestType input alignas(RTNEURAL_DEFAULT_ALIGNMENT)[num_features_in];
```
Comment (DamRsn) [inline tests/conv2d_model.h:53]: Use a vector here instead as `num_features_in` is not const (seems to be the reason for the crash on windows)

File: RTNeural/CMakeLists.txt [hunk 4]
```diff
@@ -7,6 +7,10 @@ add_library(RTNeural STATIC
     Layer.h
     conv1d/conv1d.h
     conv1d/conv1d.tpp
+    conv1d_stateless/conv1d_stateless.h
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/CMakeLists.txt:10]: Done. 

Is it normal that `conv1d_eigen`,  `conv1d_xsimd` ... are not here?
Also true for all batchnorm 1D related files (`batchnorm/batchnorm.h`).

File: RTNeural/model_loader.h [hunk 5]
```diff
@@ -509,6 +688,14 @@ namespace json_parser
                 auto batch_norm = createBatchNorm<T>(model->getNextInSize(), weights, l.at("epsilon").get<T>());
                 model->addLayer(batch_norm.release());
             }
+#if RTNEURAL_USE_EIGEN
+            else if(type == "batchnorm2d")
+            {
+                model->getNextInSize();
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/model_loader.h]: Removed, I though this function was incrementing something inside of model, but no it is `const`.

File: RTNeural/model_loader.h [hunk 6]
```diff
@@ -388,6 +503,42 @@ namespace json_parser
...
+        return true;
+    }
+
+    /** Checks that a BatchNorm2DLayer (or BatchNorm2DT) has the given dimensions. */
+    template <typename T, typename BatchNormType>
+    bool checkBatchNorm2D(const BatchNormType& batch_norm, const std::string& type, int layerDims, const nlohmann::json& weights, const bool debug)
+    {
+        if(type != "batchnorm2d")
+        {
+            debug_print("Wrong layer type! Expected: BatchNorm", debug);
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/model_loader.h]: Changed in all debug print of this function

File: RTNeural/batchnorm/batchnorm_eigen.h [hunk 7]
```diff
@@ -21,6 +21,8 @@ class BatchNorm1DLayer final : public Layer<T>
     {
         inVec = Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 1>, RTNeuralEigenAlignment>(
             input, Layer<T>::in_size, 1);
+
+        // TODO: Why not make outVec a map of out? Would avoid the copy
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/batchnorm/batchnorm_eigen.h]: Done. Also removed inVec and making a map on the fly as well.

File: RTNeural/conv2d/conv2d_eigen.tpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    int in_dilation_rate, int in_stride, bool in_valid_pad)
+    : num_filters_in(in_num_filters_in)
+    , num_filters_out(in_num_filters_out)
+    , num_features_in(in_num_features_in)
+    , kernel_size_time(in_kernel_size_time)
+    , kernel_size_feature(in_kernel_size_feature)
+    , dilation_rate(in_dilation_rate)
+    , stride(in_stride)
+    , num_features_out(Conv1DStateless<T>::computeNumFeaturesOut(in_num_features_in, in_kernel_size_feature, in_stride, in_valid_pad))
+    , receptive_field(1 + (in_kernel_size_time - 1) * in_dilation_rate)
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/conv2d/conv2d_eigen.tpp]: Done

File: RTNeural/conv2d/conv2d_eigen.tpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    , valid_pad(in_valid_pad)
+    , Layer<T>(in_num_features_in * in_num_filters_in, Conv1DStateless<T>::computeNumFeaturesOut(in_num_features_in, in_kernel_size_feature, in_stride, in_valid_pad) * in_num_filters_out)
+{
+    bias = Eigen::Vector<T, Eigen::Dynamic>::Zero(num_filters_out);
+
+    for(int i = 0; i < receptive_field; i++)
```
[the remark this answers is not in the document]
Comment (DamRsn) [reply RTNeural/conv2d/conv2d_eigen.tpp]: Yes done

File: RTNeural/conv2d/conv2d_eigen.tpp (hunk 1 above)
Comment (DamRsn) [reply RTNeural/conv2d/conv2d_eigen.tpp]: Done. Also did it for `conv1dstateless`.

File: python/conv2d.py (hunk 2 above)
Comment (DamRsn) [reply python/conv2d.py:18]: I think it's the other way around. Here the `affine == true` is tested and the non affine path needs to be tested.

In `python=`conv.py`, the first BN is affine and the second is not (`center=False, scale=False`

Comment (DamRsn) [reply python/conv2d.py:20]: Done
